src/protocolspectator.cpp

```diff
@@ -0,0 +1,378 @@
+ * with this program; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#include "otpch.h"
+
+#include "protocolgame.h"
+#include "protocolspectator.h"
+
+#include "outputmessage.h"
+
+#include "tile.h"
+#include "player.h"
+#include "chat.h"
+
+#include "configmanager.h"
+
+#include "game.h"
+
+#include "connection.h"
+#include "scheduler.h"
+#include "ban.h"
+
+extern Game g_game;
+extern ConfigManager g_config;
+extern Chat* g_chat;
+
+ProtocolSpectator::ProtocolSpectator(Connection_ptr connection):
+	ProtocolGameBase(connection),
+	client(nullptr)
+{
+
+}
+
+void ProtocolSpectator::disconnectSpectator(const std::string& message) const
+{
+	auto output = OutputMessagePool::getOutputMessage();
+	output->addByte(0x14);
+	output->addString(message);
+	send(std::move(output));
+	disconnect();
+}
+
+void ProtocolSpectator::onRecvFirstMessage(NetworkMessage& msg)
+{
+	if (g_game.getGameState() == GAME_STATE_SHUTDOWN) {
+		disconnect();
+		return;
+	}
+
+	operatingSystem = (OperatingSystem_t)msg.get<uint16_t>();
+	version = msg.get<uint16_t>();
+
+	msg.skipBytes(7); // U32 clientVersion, U8 clientType
+
+	if (!RSA_decrypt(msg)) {
+		disconnect();
+		return;
+	}
+
+	uint32_t key[4];
+	key[0] = msg.get<uint32_t>();
+	key[1] = msg.get<uint32_t>();
+	key[2] = msg.get<uint32_t>();
+	key[3] = msg.get<uint32_t>();
+	enableXTEAEncryption();
+	setXTEAKey(key);
+
+	if (operatingSystem >= CLIENTOS_OTCLIENT_LINUX) {
+		NetworkMessage opcodeMessage;
+		opcodeMessage.addByte(0x32);
+		opcodeMessage.addByte(0x00);
+		opcodeMessage.add<uint16_t>(0x00);
+		writeToOutputBuffer(opcodeMessage);
+	}
+
+	msg.skipBytes(1); // gamemaster flag
+	std::string password = msg.getString();
+	std::string characterName = msg.getString();
+
+	uint32_t timeStamp = msg.get<uint32_t>();
+	uint8_t randNumber = msg.getByte();
+	if (challengeTimestamp != timeStamp || challengeRandom != randNumber) {
+		disconnect();
+		return;
+	}
+
+	if (version < CLIENT_VERSION_MIN || version > CLIENT_VERSION_MAX) {
+		disconnectSpectator("Only clients with protocol " CLIENT_VERSION_STR " allowed!");
+		return;
+	}
+
+	if (g_game.getGameState() == GAME_STATE_STARTUP) {
+		disconnectSpectator("Gameworld is starting up. Please wait.");
+		return;
+	}
+
+	if (g_game.getGameState() == GAME_STATE_MAINTAIN) {
+		disconnectSpectator("Gameworld is under maintenance. Please re-connect in a while.");
+		return;
+	}
+
+	BanInfo banInfo;
+	if (IOBan::isIpBanned(getIP(), banInfo)) {
+		if (banInfo.reason.empty()) {
+			banInfo.reason = "(none)";
+		}
+
+		std::ostringstream ss;
+		ss << "Your IP has been banned until " << formatDateShort(banInfo.expiresAt) << " by " << banInfo.bannedBy << ".\n\nReason specified:\n" << banInfo.reason;
+		disconnectSpectator(ss.str());
+		return;
+	}
+	password.erase(password.begin()); //Erase whitespace from the front of the password string
```

Comment:
Linux does not like this if the string is empty. It compiles just fine, but it crashes the engine.

> terminate called after throwing an instance of 'std::length_error'
  what():  basic_string::_S_create
-su: line 1: 17860 Aborted                 (core dumped) ./tfs

> #10 0x0000000000a72769 in ProtocolSpectator::onRecvFirstMessage (this=0x378de88, msg=...)
    at /var/gameserver/theoxserver.com/src/protocolspectator.cpp:129